### PR TITLE
send NetworkFence requests to the leading CSI-Addons sidecar

### PR DESCRIPTION
**Pull request by nixpanic**
- util: add `NormalizeLeaseName()`
- sidecar: use leader election when CONTROLLER_SERVICE is supported
- controller: send NetworkFence requests to the leading CSI-Addons sidecar

Resolves: #422

### CODE REVIEW

File: internal/connection/connection_pool.go [hunk 1]
```diff
@@ -49,6 +51,19 @@ func (cp *ConnectionPool) Put(key string, conn *Connection) {
 	cp.pool[key] = conn
 }
 
+// Get returns the connection object corresponding to given key.
+func (cp *ConnectionPool) Get(key string) *Connection {
+	cp.rwlock.Lock()
+	defer cp.rwlock.Unlock()
```
Comment (Rakshith-R) [inline internal/connection/connection_pool.go:57]:
```suggestion
	cp.rwlock.RLock()
	defer cp.rwlock.Runlock()
```

I guess just read lock is enough here.

File: controllers/csiaddons/networkfence_controller.go [hunk 2]
```diff
@@ -298,26 +299,41 @@ func (nf *NetworkFenceInstance) removeFinalizerFromNetworkFence(ctx context.Cont
...
-			if cap.GetNetworkFence() == nil {
-				continue
-			}
+// getNetworkFenceClient returns a NetworkFenceClient that is the leader for
+// the given driver.
+// The NetworkFenceClient should only run on a CONTROLLER_SERVICE capable
+// CSI-Addons plugin, there can only be one plugin that holds the lease.
+func (r *NetworkFenceReconciler) getNetworkFenceClient(ctx context.Context, drivername string) (proto.NetworkFenceClient, error) {
+	// get the Lease for the driver
+	leaseName := util.NormalizeLeaseName(drivername) + "-csi-addons"
```
Comment (Rakshith-R) [inline controllers/csiaddons/networkfence_controller.go:308]: We need to check for leader for replication and controller reclaimspace operations too ?

**Review by Rakshith-R — CHANGES_REQUESTED** [review]

File: controllers/csiaddons/networkfence_controller.go (hunk 2 above)
Comment (nixpanic) [reply controllers/csiaddons/networkfence_controller.go:308]: yes, I want to do that in a separate PR, once this one has been reviewed/approved

**Review by Rakshith-R — COMMENTED** [review]
It looks good to me. 
just left small changes to logs.

found reference flow here https://github.com/kubernetes-csi/external-provisioner/blob/f43bea9397e4f1a08cb8b32b6b8909e57227782c/cmd/csi-provisioner/csi-provisioner.go#L649-L680 


Thanks,

**Review by Rakshith-R — APPROVED** [review]
Thanks !

**Review by yati1998 — APPROVED** [review]

**Pull request merged**